### ISSUE TITLE
External IDP Tokens stored in offline user session waste a lot of memory

### ISSUE THREAD

**Issue by sschu** [keycloak/keycloak#34092]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

identity-brokering

### Describe the bug

With an external IDP, Keycloak normally stores the FEDERATED_ID_TOKEN, FEDERATED_REFRESH_TOKEN and FEREDATED_ACCESS_TOKEN in the user session. They are used to generate an id_token_hint for the external IDP on logout (possibly refreshing the ID token before).

Offline sessions are essentially a copy of the normal user session, so they store these external tokens in the OFFLINE_USER_SESSIONS cache and in the database in the DATA column of the OFFLINE_USER_SESSION table.
However, these tokens are not used anywhere since there is no federated logout for offline sessions.

Especially if the external IDP used JWTs, this can easily mean that 80% of the memory used for storing an offline session in memory is wasted. This is probably not too critical for storage in the database but for cache memory it is.

### Version

26, nightly

### Regression

- [ ] The issue is a regression

### Expected behavior

When creating an offline session, Keycloak does not store tokens of external IDPs.

### Actual behavior

When creating an offline session, Keycloak does stores tokens of external IDPs wasting a lot of memory.

### How to Reproduce?

1. Configure an external IDP for a realm.
2. Request a token with `scope=offline_access` from that realm.
3. Check the `DATA` column in the `OFFLINE_USER_SESSION` table for FEDERATED_(ID|ACCESS|REFRESH)_TOKEN in the json document. This means the data is also stored in the OFFLINE_USER_SESSION cache.

**Comment by pedroigor** [comment]
Makes sense to me. Are you planning to work on a patch or I can take it?

I'll look at more details and propose a change if so.

**Comment by pedroigor** [comment]
Those notes should also be used by token exchange for transient sessions. In this case I think those notes are used.

**Comment by sschu** [comment]
@pedroigor I know that @mposolda worked in the same are recently, maybe he also has some thoughts. If you could give me some pointers where to ideally implement this, I can also work on it after I am back from my 1 week PTO.

**Comment by pedroigor** [comment]
I'm asking @mposolda to see what he did in the past around this. 

But looking more now on how we are storing tokens from IdPs, we also store the whole token response. In addition to those notes. Is that something you want to improve?

Can you also elaborate about why there is no federated logout for offline sessions? From an specification perspective, I understand that offline access is related to the client acting on behalf of the user without his presence. But in practice, a  people use offline sessions as regular sessions to overcome some of the issues we had around session availability and lifetime. Now we have persistent sessions enabled by default, this is no longer true but I'm not sure if we can ignore logout for offline sessions, including calling the logout endpoints at the IdP.

**Comment by mposolda** [comment]
@pedroigor @sschu The related issue, which I was recently involved, is https://github.com/keycloak/keycloak/issues/32650 .

We had the discussion with @stianst regarding that one. Until now, we created both "online" and offline session during regular login with `scope=offline_access` .  The output of the discussion is, that during regular login with `scope=offline_access`, we should not create "online" session (unless it already exists), but only the offline session. This will save memory etc, on the other hand, it may break some SSO scenarios as until now, the login with `scope=offline_access` created also the "online" session and hence future browser logins were able to re-use the SSO session. So far, it is already fixed for directGrants and clientCredentials (those are completely fine as they don't involve SSO sessions). There is still open subtask for fixing that for regular browser logins (assigned to @rmartinc, who also worked on the directGrant + clientCredentials).

This one issue is slightly related to that. As you pointed, people were using offline sessions even for regular browser logins. I am not sure what was the main motivation for it, but I guess one of the main limitation of "online" sessions was the absence of persistence. All "online" sessions basically always died during cluster restart, which was not the case for offline sessions. Now this is not the case anymore due the persistence sessions. So maybe now it is good time to change things like https://github.com/keycloak/keycloak/issues/32650 or this one?

So in shortcut, my vote is to go forward with this issue and remove the notes from offline session. We may probably do that in Keycloak 26.1.0 only, but not in 26.0.1 as it is still a slightly breaking change. That is what we agreed with @stianst also for https://github.com/keycloak/keycloak/issues/32650 . If people are pissed-off by the changes, we can possibly introduce some backwards compatibility options after Keycloak 26.1.0 release (This applies for this one as well as for https://github.com/keycloak/keycloak/issues/32650 ).

**Comment by pedroigor** [comment]
Thanks, it is more clear now the context.

> This one issue is slightly related to that. As you pointed, people were using offline sessions even for regular browser logins. I am not sure what was the main motivation for it, but I guess one of the main limitation of "online" sessions was the absence of persistence. All "online" sessions basically always died during cluster restart, which was not the case for offline sessions. Now this is not the case anymore due the persistence sessions. So maybe now it is good time to change things like #32650 or this one?

Yeah, persistence sessions change how people will now look at online and offline sessions. It was a key addition ...

I agree it is a good time to look at those issues. We can discuss more about [#32650](https://github.com/keycloak/keycloak/issues/32650#issuecomment-2416777809) but that will come with a cost in terms of UX because now clients will need to request `offline_access` in two steps (separate authorization requests) if they want to keep a browser-session open because the user needs to consent offline access. 

> 
> So in shortcut, my vote is to go forward with this issue and remove the notes from offline session. We may probably do that in Keycloak 26.1.0 only, but not in 26.0.1 as it is still a slightly breaking change. That is what we agreed with @stianst also for #32650 . If people are pissed-off by the changes, we can possibly introduce some backwards compatibility options after Keycloak 26.1.0 release (This applies for this one as well as for #32650 ).

About this particular issue ... Yeah, it will be a breaking change if we don't provide an `opt-in` switch. In addition to the logout, another thing I'm not 100% sure about removing these notes is how they will impact token exchange and the internal to external use cases.

However, it seems we are persisting the whole token response from an IdP when the configuration is set to store tokens. In this case, I think removing these notes won't make much difference as they can be obtained from the stored response?

**Comment by mposolda** [comment]
> Thanks, it is more clear now the context.
> 
> > This one issue is slightly related to that. As you pointed, people were using offline sessions even for regular browser logins. I am not sure what was the main motivation for it, but I guess one of the main limitation of "online" sessions was the absence of persistence. All "online" sessions basically always died during cluster restart, which was not the case for offline sessions. Now this is not the case anymore due the persistence sessions. So maybe now it is good time to change things like #32650 or this one?
> 
> Yeah, persistence sessions change how people will now look at online and offline sessions. It was a key addition ...
> 
> I agree it is a good time to look at those issues. We can discuss more about [#32650](https://github.com/keycloak/keycloak/issues/32650#issuecomment-2416777809) but that will come with a cost in terms of UX because now clients will need to request `offline_access` in two steps (separate authorization requests) if they want to keep a browser-session open because the user needs to consent offline access.

Yes, exactly. We discussed that with @stianst and so far agreement was that we should rather go that path even if it has some potential impact on the scenarios like you mentioned (you can see the comments in the #32650 etc). Feel free to comment as well if you think we should revisit the decision. ATM right now I am more inclined to change it as it has some additional flexibility (currently you have always online+offline session even if you don't need online session if you don't care at all about SSO login).

> 
> > So in shortcut, my vote is to go forward with this issue and remove the notes from offline session. We may probably do that in Keycloak 26.1.0 only, but not in 26.0.1 as it is still a slightly breaking change. That is what we agreed with @stianst also for #32650 . If people are pissed-off by the changes, we can possibly introduce some backwards compatibility options after Keycloak 26.1.0 release (This applies for this one as well as for #32650 ).
> 
> About this particular issue ... Yeah, it will be a breaking change if we don't provide an `opt-in` switch. In addition to the logout, another thing I'm not 100% sure about removing these notes is how they will impact token exchange and the internal to external use cases.
> 
> However, it seems we are persisting the whole token response from an IdP when the configuration is set to store tokens. In this case, I think removing these notes won't make much difference as they can be obtained from the stored response?

I am not sure how much we need to care about the compatibility for token exchange and internal-to-external use-case as it is a "preview" feature and hence the compatibility is not guaranteed anyway? And we have `Store tokens` option for the people, who need it as you mentioned.

**Comment by pedroigor** [comment]
Ok, thanks. Sounds good to me.

@sschu I'll take a look at it and try to get some pointers. I would like if we only need to remove the references where those notes are set to an offline session as well as check how it will impact/change in token exchange.

**Comment by pedroigor** [comment]
@sschu About what I have discussed with Marek:

> However, it seems we are persisting the whole token response from an IdP when the configuration is set to store tokens. In this case, I think removing these notes won't make much difference as they can be obtained from the stored response?

Do you think we can just remove references where those notes are being set and rely on the token response that is stored together with the federated identity whenever we need to parse tokens from the IdP (e.g.: exchange)?

**Comment by sschu** [comment]
@pedroigor It looks like the token used for federated logout for a normal session is the one stored in the session note, see https://github.com/keycloak/keycloak/blob/bbb83236f5251e3c4718eac63ab0e640a80913f1/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L170 

So we cannot remove it completely. I would probably rather remove it when creating the offline session, maybe here
https://github.com/keycloak/keycloak/blob/9454c01d8878103e75b06ecad6b14830abaf119a/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java#L789

and/or here
https://github.com/keycloak/keycloak/blob/9454c01d8878103e75b06ecad6b14830abaf119a/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java#L664

and/or here
https://github.com/keycloak/keycloak/blob/ba861fc5d7aa8e90be29cc704c482123d5d913bb/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteUserSessionProvider.java#L219

WDYT?

**Comment by sschu** [comment]
@rmartinc Thanks for your feedback. On the IDP, there is a setting "Store tokens" already. Let me check if this has any effect...

**Comment by sschu** [comment]
@rmartinc To me, it looks like this setting controls if external tokens are stored in the `FederatedIdentity`, see https://github.com/keycloak/keycloak/blob/731274f39e522d6a852f630796aac4452912683b/services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java#L499 . They seem to be always stored as a session note, currently. 

The external tokens stored in the session note seem to be used for a) federated logout and b) for token exchange if the `Store tokens` setting is not enabled (see 
https://github.com/keycloak/keycloak/blob/bbb83236f5251e3c4718eac63ab0e640a80913f1/services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java#L199-L211)
I don't really know why there is this fallback. I would have expected it to be perfectly fine if you cannot exchange tokens if you did not enable `Store tokens`.

**Comment by pedroigor** [comment]
@sschu Sounds like a good idea to only remove when creating the offline sessions. 

However, I was wondering if we can instead rely on the token stored within the federated entity and always parse it (e.g.: once per session and on-demand). My point is that we have the information stored in the session and in the federated identity so that we can have a single source of truth for tokens issued by an IdP from the federated identity.

**Comment by stianst** [comment]
Haven't read through all comments, but in general I would say an offline token should have no link whatsoever to the federated session/token.

As examples:

* It should not be possible to do a internal to external token exchange with a offline token
* Logout of an offline session should not trigger a logout to the IdP
* It should not be possible to get federated tokens from session in mappers, or any other way

The fact that we are storing these tokens in regular sessions may not be ideal either, but I would say that is a separate issue. @sschu also mentioned that we are refreshing the ID token before logging out the federated session, but that is wasteful as far as I know the spec says an IdP should accept an expired ID token.
